torch/autograd/graph.py

```diff
@@ -112,21 +113,24 @@ class save_on_cpu(saved_tensors_hooks):
         >>> # all intermediary tensors are released (deleted) after the call to backward
 
     """
+    # NOTE: when pin_memory=True, save_on_cpu does asynchronous copy when moving
```

Comment:
This properly explains why this is safe wrt the user but not why this function is safe.
In particular, we want to mention that we rely on the following facts:
- Any access to the packed Tensor within the SavedVariable is guaranteed to be done through the "unpack" hook.
- The autograd engine is guaranteed to properly restore the current stream that was used during forward before unpacking the SavedVariables for backward.

But there are two points that we are missing I think:
- What is ensuring that when the user accesses this from outside the backward via `grad_fn.saved_foo` that the current stream is properly set? 
- What ensures that the saved cpu tensor is kept alive long enough for the copy back to the gpu to happen?

Reply:
> What is ensuring that when the user accesses this from outside the backward via grad_fn.saved_foo that the current stream is properly set?

Do you mean accessing these tensors with `grad_fn.saved_tensors` outside of backward? Is this intended to be a supported feature, as I thought we were assuming this packed tensor should never be accessed by the user (and there's no way to do so)? 

> What ensures that the saved cpu tensor is kept alive long enough for the copy back to the gpu to happen?

I think as you mentioned https://github.com/pytorch/pytorch/blob/3c2ae2b47cc859847e7915a406a7484401cba35e/aten/src/ATen/native/cuda/Copy.cu#L227 should prevent against deallocation happening?

Reply:
> Do you mean accessing these tensors with grad_fn.saved_tensors outside of backward? Is this intended to be a supported feature, as I thought we were assuming this packed tensor should never be accessed by the user (and there's no way to do so)?

Yes this is a supported feature. mostly used by visualization tools. But still. 
I think this can be fixed by making sure that the streaming is done properly during unpacking.

> should prevent against deallocation happening?

Yes. We should verify that this is what it is doing and add that to the list of things that we rely on.

> In addition, would there be a difference in the semantics with regard to copy back to GPU?

Yes because now the CPU memory might not be filed when we reach the unpack function. So the unpack function needs to be extra careful that it won't try to read it until it is ready (mostly by ordering properly wrt to the packing's ambient stream I think).